### ISSUE TITLE
Can't install >v0.1.3 KubeFlow to Microk8s

### ISSUE THREAD

**Issue by hyzhak** [kubeflow/kubeflow#1854]
I've followed [instruction of Microk8s](https://www.kubeflow.org/docs/started/getting-started-multipass/) and have installed `0.1.3` successfully.

But when I'm trying to install newer versions `0.2.7` or `0.3.1` they won't work. 

I've got:

> $ VERSION=v0.3.1 /kubeflow/install-kubeflow.sh                                                     
> namespace/kubeflow created                                                                                             
> INFO Using context "microk8s" from kubeconfig file "/snap/microk8s/current/client.config"                              
> INFO Creating environment "default" with namespace "default", pointing to cluster at address "http://127.0.0.1:8080"   
> INFO Generating ksonnet-lib data at path '/home/multipass/my-kubeflow/lib/v1.12.0'                                     
> INFO Retrieved 38 files                                                                                                
> INFO Retrieved 9 files                                                                                                 
> ERROR resolve registry library: GET https://api.github.com/repos/kubeflow/kubeflow/contents/kubeflow/tf-job?ref=c456ceec089675e7f5c4a7494fb0f7ce875fe062: 404 Not Found []

> $ VERSION=v0.2.7 /kubeflow/install-kubeflow.sh                                                     
> namespace/kubeflow created                                                                                             
> INFO Using context "microk8s" from kubeconfig file "/snap/microk8s/current/client.config"                              
> INFO Creating environment "default" with namespace "default", pointing to cluster at address "http://127.0.0.1:8080"   
> INFO Generating ksonnet-lib data at path '/home/multipass/my-kubeflow/lib/v1.12.0'                                     
> INFO Retrieved 33 files                                                                                                
> INFO Retrieved 5 files                                                                                                 
> ERROR resolve registry library: GET https://api.github.com/repos/kubeflow/kubeflow/contents/kubeflow/tf-job?ref=4ffd0a5aa97cea1eddcab185fd33da08e9efad8a: 404 Not Found []

Is `Microk8s` still in business or I should switch to `Minikube` to get newer `Kubeflow`?

**Comment by jlewi** [comment]
Are you following?
https://www.kubeflow.org/docs/started/getting-started-multipass/

I think the install_kubeflow.sh script is out of date. The TFJob package was removed (prior to 0.2 I think).

Once you have a K8s cluster (e.g. MicroK8s) and can access the master you can just follow the generic instructions
https://www.kubeflow.org/docs/started/getting-started/

To deploy Kubeflow on that cluster.

@carmine Is the install_kubeflow script shipped as part of MicroK8s? Should we update the docs?

**Comment by carmine** [comment]
Will update the instructions this weekend based on reproducing / resolving the issue. Not sure if this is related to the snapd/docker issue or not, which has been resolved in the edge channel, but not in stable channel yet.

**Comment by hyzhak** [comment]
Thanks @jlewi and @carmine! Yes, I had thought to try to setup kubeflow manually there. But recently I've found another problem - `multipass` which is used for `microk8s` setup. It doesn't allow to get access to `kubeflow` services (Jupyter and other) from local network https://github.com/CanonicalLtd/multipass/issues/118. So it won't work in my case (small local DL server for rnd) at least until it will be solved. 

But could it make sense to use `kubeflow` and `microk8s` without `multipass`?

It is just digression and new problem isn't related to the current issue. So maybe this problem requires to build extra issue here?

**Comment by ktsakalozos** [comment]
Hi @hyzhak ,

`microk8s` is still in business. Yes you can use `microk8s` outside `multipass` as long as you are on ubuntu, just do a `sudo snap install microk8s --classic`. The kubeflow installation instructions give you `v0.1.3` because we are waiting on a bug fix to be released(*). 

If you want to deploy a v0.2 kubeflow you will need to use a `jupyterHubImage` that lives in `kjackal/jupyterhub-k8s:v0.9.2` essentially this means you have to `ks param set kubeflow-core jupyterHubImage kjackal/jupyterhub-k8s:v0.9.2`. If you do not want to wait for the above fix to be released, here is the `v0.2.0` deployment script you have to run after having ks and microk8s installed.

```
#!/usr/bin/env bash
set -eux

sudo snap alias microk8s.kubectl kubectl
mkdir -p ~/.kube
microk8s.config > ~/.kube/config
microk8s.enable dns dashboard storage
sudo iptables -P FORWARD ACCEPT

KUBEFLOW_VERSION=${KUBEFLOW_VERSION:-v0.2.2}
wget https://raw.githubusercontent.com/kubeflow/kubeflow/v${KUBEFLOW_VERSION}/scripts/deploy.sh
cat ./deploy.sh | awk '{if ($n ~ "# Apply the components generated") {print "ks param set kubeflow-core jupyterHubImage kjackal/jupyterhub-k8s:v0.9.2"; print $n} else print $n}' > deploy-kf.sh
bash ./deploy-kf.sh
```

* The fix on https://github.com/snapcore/snapd/pull/5739 is already merged into the microk8s stack and is about to be released.

**Comment by ktsakalozos** [comment]
I see the `v.0.2.7` installation process is the same as the on in `v0.2.2` so I would like to think the above script works. For `v0.3.1` the installation scripts have changed so I do not know yet where to put the `ks param set kubeflow-core jupyterHubImage kjackal/jupyterhub-k8s:v0.9.2`.

**Comment by hyzhak** [comment]
@ktsakalozos ok thanks! So, do you have plan to adapt to `v0.3.x` and I could wait for a few days or it would better to use `v0.2.x`?

**Comment by ktsakalozos** [comment]
@hyzhak , @carmine with the https://github.com/snapcore/snapd/pull/5739 fix in place the kubeflow installation in microk8s does not need any special configuration (this is not surprising since microk8s is about to get the cncf certification https://github.com/cncf/k8s-conformance/pull/360). Here is how I got `v0.3.1` kubeflow running on a t2.xlarge instance on aws.

```
# For now the aforementioned fix is available on the snap core of the edge channel.
# The following command will not be needed as soon as core is released on the stable channel
sudo snap refresh core --edge

# Install microk8s
sudo snap install microk8s --classic

# Ensure we have kubectl in our path and we place the kubeconfig is the default location
sudo snap alias microk8s.kubectl kubectl
mkdir -p ~/.kube
microk8s.config > ~/.kube/config

# Allow intra-pod communication
sudo iptables -P FORWARD ACCEPT

# Enable dns dashboard and default storage
microk8s.enable dns dashboard storage

# Install ksonnet and add ks in our path
wget https://github.com/ksonnet/ksonnet/releases/download/v0.13.0/ks_0.13.0_linux_amd64.tar.gz
tar -zxvf ./ks_0.13.0_linux_amd64.tar.gz
sudo cp ks_0.13.0_linux_amd64/ks /usr/bin

# Set the two environment variables used in the getting started instructions 
# https://www.kubeflow.org/docs/started/getting-started/
export KUBEFLOW_SRC=$(pwd)/kfsrc
export KFAPP=$KUBEFLOW_SRC/kfapp

# Follow the installation steps
mkdir ${KUBEFLOW_SRC}
cd ${KUBEFLOW_SRC}
export KUBEFLOW_TAG=v0.3.1
curl https://raw.githubusercontent.com/kubeflow/kubeflow/${KUBEFLOW_TAG}/scripts/download.sh | bash
${KUBEFLOW_SRC}/scripts/kfctl.sh init ${KFAPP} --platform none
cd ${KFAPP}
${KUBEFLOW_SRC}/scripts/kfctl.sh generate k8s
${KUBEFLOW_SRC}/scripts/kfctl.sh apply k8s

# Wait while checking the installation progress
watch kubectl get all --all-namespaces
```

This is the final state: http://paste.ubuntu.com/p/GVpjRwzqhj/

**Comment by leigh-johnson** [comment]
I walked through the setup currently documented @ [MicroK8s for Kubeflow](https://www.kubeflow.org/docs/started/getting-started-multipass/) and reproduced the issues in the OP using:

* Host machine: MacOS 10.13.6
* Multipass 2018.6.1
* Ubuntu 18.04 via dockerhub
* Kubernetes v1.12.2 (via microk8s 1.12/beta release channel)

After I updated `install_kubeflow.sh` to follow the instructions documented @ [Getting Started With Kubeflow](https://www.kubeflow.org/docs/started/getting-started/), I was able to get this setup working. Let me know if you'd like to incorporate upstream. @carmine 

I published my revisions @ https://github.com/leigh-johnson/kubeflow-microk8s-demo as part of a workshop I'm running.

Extra note: the `ksonnet` release binaries I tested (`v0.11.1` & `v0.13.0`) panic'd with Kubernetes v1.12.*. I built ksonnet from master to [grab a patch for this issue](https://github.com/shomron/ksonnet/commit/8ce41f5e959bd1e9f84e08de0c1ee879ba21cdae), but this could introduce other unexpected issues for existing setups.

**Comment by jlewi** [comment]
@carmine WDYT should we incoporate @leigh-johnson's fixes?

**Comment by carmine** [comment]
Will look at the patch by Monday 12/31. I’ve experimented with some additional scripting, including exposing K8s Dashboard. Will update instructions accordingly.

**Comment by carmine** [comment]
Incorporating changes now and will have a PR next 1-2 days.

**Comment by carmine** [comment]
Have created PR https://github.com/kubeflow/website/pull/391. Will close this issue when that has been merged. The instructions should clarify that users can use microk8s without multipass if they are on a Linux machine. Additionally, the instructions have been updated to leverage kfctl.sh and download.sh.  The serviceType for ambassador has been changed to NodePort to simplify access from outside the machine (e.g. if using multipass / VM).

**Comment by carmine** [comment]
The updated instructions have hit the website, and the current scripts install the latest ksonnet (0.13.1) / kubeflow (0.3.5). Closing this issue.
